storage: recompute MVCCStats during consistency checks

tbg opened the pull request:
This lays the foundation for planned work in which the consistency checker can
trigger a stats recomputation should the (consistently replicated) stats be
found to be incorrect. In light of #20996, this is evidently something we need.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21207)
<!-- Reviewable:end -->

nvanbenschoten commented:
Reviewed 2 of 3 files at r1.
Review status: 2 of 3 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_consistency.go, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207#-L1xhjUUYtZKWGoRa30i:-L1xhjUUYtZKWGoRa30j:b-hhf3ge) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f5020e82ecf132ff230cbe77fd98860ed780a6/pkg/storage/replica_consistency.go#L27)):*
> ```Go
> 	"time"
> 
> 	"github.com/cockroachdb/cockroach/pkg/storage/engine/enginepb"
> ```

move down ;)

---

*[pkg/storage/replica_consistency.go, line 325 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207#-L1xi17ex_bLI_bp8CpW:-L1xi17ex_bLI_bp8CpX:bcjs70f) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f5020e82ecf132ff230cbe77fd98860ed780a6/pkg/storage/replica_consistency.go#L325)):*
> ```Go
> }
> 
> type replicaHash struct {
> ```

This might be a bit large to be passing around by value. What do you think?

---

*[pkg/storage/replica_consistency.go, line 358 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207#-L1xiyOZIJtJM1_BpNA8:-L1xiyOZIJtJM1_BpNA9:b-31fdah) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f5020e82ecf132ff230cbe77fd98860ed780a6/pkg/storage/replica_consistency.go#L358)):*
> k:v

That's... exotic

---

*[pkg/storage/replica_consistency.go, line 360 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207#-L1xjCKhRlUQGyDZdbhq:-L1xjCKhRlUQGyDZdbhr:b-i36ldv) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f5020e82ecf132ff230cbe77fd98860ed780a6/pkg/storage/replica_consistency.go#L360)):*
> ```Go
> 			// Add (a copy of) the k:v into the debug message.
> 			snapshot.KV = append(snapshot.KV, roachpb.RaftSnapshotData_KeyValue{
> 				Key:       append(roachpb.Key(nil), unsafeKey.Key...),
> ```

Should we use a `bufalloc.ByteAllocator` for this, like we do in `MVCCIterate`?

---

*[pkg/storage/replica_consistency.go, line 414 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207#-L1xl_njX7-METczI538:-L1xl_njX7-METczI539:bwlv7ty) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f5020e82ecf132ff230cbe77fd98860ed780a6/pkg/storage/replica_consistency.go#L414)):*
> ```Go
> 	}
> 
> 	result.RecomputedMS.AgeTo(result.PersistedMS.LastUpdateNanos)
> ```

Could you add a comment for this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207)*
<!-- Sent from Reviewable.io -->

tbg committed:
sql: "k:v" -> "kv"

Release note: None

tschottdorf commented:
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/storage/replica_consistency.go, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207#-L1xhjUUYtZKWGoRa30i:-L1xmqW9YM_kaB3pS_ga:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f5020e82ecf132ff230cbe77fd98860ed780a6/pkg/storage/replica_consistency.go#L27)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

move down ;)
</blockquote></details>

Done.

---

*[pkg/storage/replica_consistency.go, line 325 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207#-L1xi17ex_bLI_bp8CpW:-L1xnKgn_Xo0Izk7wzak:b-cql58b) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f5020e82ecf132ff230cbe77fd98860ed780a6/pkg/storage/replica_consistency.go#L325)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This might be a bit large to be passing around by value. What do you think?
</blockquote></details>

Yeah, good point. Changed.

---

*[pkg/storage/replica_consistency.go, line 358 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207#-L1xiyOZIJtJM1_BpNA8:-L1xnI2ogJi-DoPBA7JM:b-k3sc6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f5020e82ecf132ff230cbe77fd98860ed780a6/pkg/storage/replica_consistency.go#L358)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> k:v

That's... exotic
</blockquote></details>

Fixed (throughout the project).

---

*[pkg/storage/replica_consistency.go, line 360 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207#-L1xjCKhRlUQGyDZdbhq:-L1xoDBPRCkEYvlnnCue:b-jn1moz) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f5020e82ecf132ff230cbe77fd98860ed780a6/pkg/storage/replica_consistency.go#L360)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we use a `bufalloc.ByteAllocator` for this, like we do in `MVCCIterate`?
</blockquote></details>

I added that, though I anticipate that once we allow for larger ranges, we're going to have to throw this under the bus (and use Merkle trees).

---

*[pkg/storage/replica_consistency.go, line 414 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207#-L1xl_njX7-METczI538:-L1xo_cB2OBr8KiP1xxR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f5020e82ecf132ff230cbe77fd98860ed780a6/pkg/storage/replica_consistency.go#L414)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Could you add a comment for this?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Reviewed 1 of 4 files at r2.
Review status: 1 of 5 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_consistency.go, line 325 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207#-L1xi17ex_bLI_bp8CpW:-L1xt53xB2FVvBN95ZnB:bbx5sjy) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f5020e82ecf132ff230cbe77fd98860ed780a6/pkg/storage/replica_consistency.go#L325)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Yeah, good point. Changed.
</blockquote></details>

Oh sorry, I like the array, I meant `replicaHash` itself.

---

*[pkg/storage/replica_consistency.go, line 364 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207#-L1xtU1YO-ofg1jmdpcJ:-L1xtU1YO-ofg1jmdpcK:byo6w8n) ([raw file](https://github.com/cockroachdb/cockroach/blob/41aeea215c561a3bacae9fcd43a1f3286e2de3b0/pkg/storage/replica_consistency.go#L364)):*
> ```Go
> 				Timestamp: unsafeKey.Timestamp,
> 			}
> 			alloc, kv.Key = alloc.Copy(unsafeKey.Key, len(unsafeValue))
> ```

I think we want `0` for extra cap, which will be added to `kv.Key` and won't be used for `kv.Value`. Even if it was though, reserving more than we need is probably a pessimization because we'll waste more space.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 1 of 5 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_consistency.go, line 325 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207#-L1xi17ex_bLI_bp8CpW:-L1xz-jZi4ztgcE-bLii:b-3j4jod) ([raw file](https://github.com/cockroachdb/cockroach/blob/41f5020e82ecf132ff230cbe77fd98860ed780a6/pkg/storage/replica_consistency.go#L325)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Oh sorry, I like the array, I meant `replicaHash` itself.
</blockquote></details>

Gotcha, done.

---

*[pkg/storage/replica_consistency.go, line 364 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207#-L1xtU1YO-ofg1jmdpcJ:-L1xz1om2ZPQ_gxG7ECp:b-mwpkt9) ([raw file](https://github.com/cockroachdb/cockroach/blob/41aeea215c561a3bacae9fcd43a1f3286e2de3b0/pkg/storage/replica_consistency.go#L364)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I think we want `0` for extra cap, which will be added to `kv.Key` and won't be used for `kv.Value`. Even if it was though, reserving more than we need is probably a pessimization because we'll waste more space.
</blockquote></details>

I misunderstood the semantics. Fixed!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 1 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/storage/replica_consistency.go, line 307 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207#-L1y1SUoLzHdZpx0EWyP:-L1y1SUpAEzl7foi5jW8:b-6mr00e) ([raw file](https://github.com/cockroachdb/cockroach/blob/03809da073468c1f102ddb5c6906006c2bc2a28c/pkg/storage/replica_consistency.go#L307)):*
> ```Go
> // checksum, and sends out a notification.
> func (r *Replica) computeChecksumDone(
> 	ctx context.Context, id uuid.UUID, result replicaHash, snapshot *roachpb.RaftSnapshotData,
> ```

You could make this a pointer too. Your call.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207#-:-L1y1VMMG66aaGJq60Da:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR!

---

Review status: 1 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/storage/replica_consistency.go, line 307 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207#-L1y1SUoLzHdZpx0EWyP:-L1y1YwGXYsdukqFHhII:b4xizox) ([raw file](https://github.com/cockroachdb/cockroach/blob/03809da073468c1f102ddb5c6906006c2bc2a28c/pkg/storage/replica_consistency.go#L307)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

You could make this a pointer too. Your call.
</blockquote></details>

We store a result in the checksum map either way, so I intentionally am passing a valued-result in here. Possibly I'll change it down the road, but for now I'll leave as-is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207#-:-L1y1kp8QZoD2WO58ds-:b-814g8k)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: recompute MVCCStats during consistency checks

This lays the foundation for planned work in which the consistency checker can
trigger a stats recomputation should the (consistently replicated) stats be
found to be incorrect. In light of #20996, this is evidently something we need.

Release note: None

bdarnell commented:
Doing this in the consistency checker (where the work will be repeated on every replica) feels odd to me. But I don't know of a better place to put this (without creating a new queue for the purpose).

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21207#-:-L1yx6yd3t2udPpLqIhQ:b-tkms2p)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
@bdarnell that's a good point. In principle we can do this work on the lease holder only (but then we better be pretty sure that the Go and C++ status computations are identical, though this is well-tested and the code is also well-contained) in the future.

tschottdorf commented:
(and as for the placing it in the consistency checker in general, I think it's the right place in the long-term -- a few releases down the road, we won't have inconsistencies in our stats that we know about, and so having the consistency checker do a recomputation and emit a sentry event when a divergence is detected will be useful and fits the general theme of this queue).
